include/xtensor-r/roptional.hpp

```diff
@@ -0,0 +1,268 @@
+        using storage_type = xoptional_assembly_storage<value_storage_type&, flag_storage_type&>;
+        using temporary_type = roptional_assembly<T>;
+    };
+
+    template <class T>
+    struct xiterable_inner_types<roptional_assembly<T>>
+    {
+        using assembly_type = roptional_assembly<T>;
+        using inner_shape_type = typename rarray<T>::inner_shape_type;
+        using stepper = xoptional_assembly_stepper<assembly_type, false>;
+        using const_stepper = xoptional_assembly_stepper<assembly_type, true>;
+    };
+
+    template <class T>
+    class roptional_assembly : public xoptional_assembly_base<roptional_assembly<T>>,
+                               public xcontainer_semantic<roptional_assembly<T>>
+    {
+    public:
+
+        using self_type = roptional_assembly<T>;
+        using base_type = xoptional_assembly_base<self_type>;
+        using semantic_base = xcontainer_semantic<self_type>;
+
+        using storage_type = typename base_type::storage_type;
+        using assembly_type = roptional_assembly<T>;
+
+        explicit roptional_assembly(SEXP exp);
+
+        roptional_assembly(const roptional_assembly&);
+        roptional_assembly& operator=(const roptional_assembly&);
+        roptional_assembly(roptional_assembly&&);
+        roptional_assembly& operator=(roptional_assembly&&);
+
+        template <class E>
+        roptional_assembly(const xexpression<E>& e);
+
+        template <class E>
+        roptional_assembly& operator=(const xexpression<E>& e);
+
+    private:
+
+        auto& value_impl() noexcept;
+        const auto& value_impl() const noexcept;
+        auto& has_value_impl() noexcept;
+        const auto& has_value_impl() const noexcept;
+        auto& storage_impl() noexcept;
+        const auto& storage_impl() const noexcept;
+
+        rarray<T> m_value;
+        xfunctor_adaptor<rna_proxy_functor<T>, rarray<T>&> m_flag;
+        storage_type m_storage_proxy;
+
+        friend xoptional_assembly_base<roptional_assembly<T>>;
+    };
+
+    /*************************************
+     * roptional_assembly implementation *
+     *************************************/
+
+    template <class T>
+    inline roptional_assembly<T>::roptional_assembly(SEXP exp)
+        : m_value(exp), m_flag(m_value), m_storage_proxy(m_value.storage(), m_flag)
+    {
+    }
+
+    template <class T>
+    inline roptional_assembly<T>::roptional_assembly(const roptional_assembly& rhs)
+        : m_value(rhs.m_value), m_flag(rhs.m_flag), m_storage_proxy(m_value.storage(), m_flag)
+    {
+    }
+
+    template <class T>
+    inline auto roptional_assembly<T>::operator=(const self_type& rhs) -> self_type&
+    {
+        base_type::operator=(rhs);
+        m_value = rhs.m_value;
+        return *this;
+    }
+
+    template <class T>
+    inline roptional_assembly<T>::roptional_assembly(self_type&& rhs)
+        : m_value(rhs.m_value), m_flag(rhs.m_flag), m_storage_proxy(m_value.storage(), m_flag)
```

Comment:
use m_value for m_flag so that reference stays valid. also above.

Reply:
and move if possible :)

Reply:
fixed!